### ISSUE TITLE
[API] Create endpoint to confirm a pending follower request

### ISSUE THREAD

**Issue by MrPetovan** [friendica/friendica#7887]
Part of #6739

What about Mastodon's API endpoints:
- Approve: https://docs.joinmastodon.org/api/rest/follow-requests/#post-api-v1-follow-requests-id-authorize
- Reject: https://docs.joinmastodon.org/api/rest/follow-requests/#post-api-v1-follow-requests-id-reject

cc @stom79

**Comment by stom79** [comment]
That would be perfect :+1:

**Comment by MrPetovan** [comment]
Curious question, how do you decide which endpoint to use in Fedilab? Is there a matrix by target software? I'm surprised you welcome combining a Twitter endpoint to a couple of Mastodon's endpoint to interact with Friendica.

**Comment by stom79** [comment]
In fact, I am not really choosing which endpoints to use because I want to use them all :) 
For pointing what endpoints I would like to see with Friendica, it's first for implementing most appreciated features.
The app allows to follow other instances, it's only a public api call. But as soon as a user interacts, the app needs to have the ability to federate this message and the attached account to the instance. That's why I am deeply interested to the search feature.
Also, this is how the app works for cross-account actions. With that feature, people having an account on Mastodon and Friendica can boost, fav, reply, follow with the other account without switching. 
And yes, Friendica has its dedicated management inside the app, so I can add specific features or layouts :) 
But the Friendica community using Fedilab tells me also what they would like to have.

**Comment by annando** [comment]
@stom79 BTW: where can I place some feature requests? Currently quoted posts aren't displayed well.

**Comment by MrPetovan** [comment]
I started working on this instead of the custom profile fields feature.

**Comment by stom79** [comment]
> @stom79 BTW: where can I place some feature requests? Currently quoted posts aren't displayed well.

Could you provide me examples so I can compare with the web app and work on that issue. Otherwise, issues for Fedilab are here: https://framagit.org/tom79/fedilab/issues

**Comment by annando** [comment]
The quotes are described here: https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/intro-to-tweet-json#quotetweet

Additionally to the regular text they have got some `quoted_status` field. And it looks like this: https://twitter.com/hbhammel/status/1201553619440545793

The `quoted_status` contains everything that the original post has got as well, including attachments.

**Comment by stom79** [comment]
OK. I will work on it: https://framagit.org/tom79/fedilab/issues/367

**Comment by stom79** [comment]
Also, could we agree to use the same URL for endpoints that will come with the Mastodon API? I mean they will start by /api/(v1|v2)/same_endpoint_name (and parameters)?

That will really help in future when some endpoints will be migrated to the Mastodon API. On my end, it's really not a problem to mix api for a same software (I already do it for Pixelfed). 
Currently, I am working with the Twitter api for implementing some missing endpoints for Friendica.

**Comment by annando** [comment]
I would prefer implementing new API calls with Mastodon-like endpoints. But I'm unsure if we will have it done for the current release that is planned for the end of the year. We currently do have some ugly behaviour with reshares that are having priority.

**Comment by MrPetovan** [comment]
> Also, could we agree to use the same URL for endpoints that will come with the Mastodon API? I mean they will start by /api/(v1|v2)/same_endpoint_name (and parameters)?

Yes, that's what we meant when we say we implement Mastodon/Twitter's API endpoint. We use the same URL, same parameter and same return values.

**Comment by stom79** [comment]
> The quotes are described here: https://developer.twitter.com/en/docs/tweets/data-dictionary/overview/intro-to-tweet-json#quotetweet
> 
> Additionally to the regular text they have got some `quoted_status` field. And it looks like this: https://twitter.com/hbhammel/status/1201553619440545793
> 
> The `quoted_status` contains everything that the original post has got as well, including attachments.

OK. I checked and I would need to have them attached to the status object. I can't do special calls for each cardview that would be too costly. The html content will not help for apps as they don't run webview.

Also, I noticed that the api returns some empty links like, I guess for a media gallery? 
`<a href="https://libranet.de/photos/USERNAME/image/17410654725de40af16e4f2179729508">￼</a>`
But it can't be used directly.

**Comment by annando** [comment]
The "quoted_status" is attached to the original "status" object and contains the whole stuff (the HTML, the user array and so on). There is no need for refetching anything. Isn't it possible to "simply" attach it with some `<blockquote>...</blockquote>` surrounding it?

**Comment by stom79** [comment]
Not so simple to mix (transform html) ie via friendica_html
I tried but now reblogs (shares) are affected and are ugly (lot of spaces). That would need more logic. I think that the best would be to use a card like Mastodon https://docs.joinmastodon.org/api/entities/#card

**Comment by MrPetovan** [comment]
Can you please move over this discussion to Fedilab's issue tracker?

**Comment by MrPetovan** [comment]
First part is done, now I need to provide the endpoint to accept/reject a follow request. This should be faster as I won’t have to rewrite the entire API documentation again 😅

**Comment by annando** [comment]
You have to ensure to not confirm requests from native DFRN. Only AP, Diaspora and DFRN via AP can be confirmed via API.

**Comment by MrPetovan** [comment]
Thanks for the tip. Should I exclude DFRN contact follow requests from the GET endpoint list?

**Comment by annando** [comment]
It's slightly more complicated than simply DFRN contacts. Have a look at the "protocol" field as well. When this is filled (with AP), then it can be confirmed as well.

**Comment by MrPetovan** [comment]
Ok thanks.

**Comment by annando** [comment]
Do you know the function to accept requests or should I help searching for it?

BTW: If you want to add some endpoint to follow someone as well, have a look at `Contact::createFromProbe`.

**Comment by MrPetovan** [comment]
No, I'm going to limit myself to the task at hand. I was planning on following the web modules.

**Comment by MrPetovan** [comment]
With #7930 this issue should now be complete. @stom79 can you please confirm the endpoints are working as expected on `libranet.de`?

**Comment by AlfredSK** [comment]
Wait. I haven't pulled that one yet. 😁

**Comment by MrPetovan** [comment]
I was assuming that by the time @stom79 would have read my message and found time to test, you would already have.

**Comment by AlfredSK** [comment]
Now I'm ready.

**Comment by MrPetovan** [comment]
Any news @stom79 ?

**Comment by stom79** [comment]
Currently, this call: ` https://libranet.de/api/v1/follow_requests/%s/authorize` (**%s** string id for an account), returns:
```html
<html>
    	<head>
    		<title>Not found</title>
    	</head>
    	<body>
    		<h1>Not found</h1>
    		<p></p>
    	</body>
    </html>
```
Could it be related to an error with the id %s?

**Comment by stom79** [comment]
OK. Please ignore my previous message. I was not using the new endpoint for getting follow requests and user ids were different.
It perfectly works when fetching pending follow requests with the new endpoint + post authorize.

**Comment by annando** [comment]
@MrPetovan possibly we should always publish the public contact id? I guess this avoids some confusion.

**Comment by MrPetovan** [comment]
> @MrPetovan possibly we should always publish the public contact id? I guess this avoids some confusion.

I don't think so, the endpoint is `follow_requests` so it returns a follow request ID.

**Comment by MrPetovan** [comment]
Closed by #7930
